src/Auth.js

```diff
@@ -332,6 +347,163 @@ const createSession = function (
+  const providers = Object.keys(authData);
+  const query = providers
+    .reduce((memo, provider) => {
+      if (!authData[provider] || (authData && !authData[provider].id)) {
+        return memo;
+      }
+      const queryKey = `authData.${provider}.id`;
+      const query = {};
+      query[queryKey] = authData[provider].id;
+      memo.push(query);
+      return memo;
+    }, [])
+    .filter(q => {
+      return typeof q !== 'undefined';
+    });
+
+  let findPromise = Promise.resolve([]);
+  if (query.length > 0) {
+    findPromise = config.database.find('_User', { $or: query }, {});
+  }
+
+  return findPromise;
+};
+
+const hasMutatedAuthData = (authData, userAuthData, config) => {
+  if (!userAuthData) return { hasMutatedAuthData: true, mutatedAuthData: authData };
+  const mutatedAuthData = {};
+  Object.keys(authData).forEach(provider => {
+    // Anonymous provider is not handled this way
+    if (provider === 'anonymous') return;
+    const providerData = authData[provider];
+    const userProviderAuthData = userAuthData[provider];
+    if (!_.isEqual(providerData, userProviderAuthData) || config.auth[provider].alwaysValidate) {
+      mutatedAuthData[provider] = providerData;
+    }
+  });
+  const hasMutatedAuthData = Object.keys(mutatedAuthData).length !== 0;
+  return { hasMutatedAuthData, mutatedAuthData };
+};
+
+const checkRequiredProviders = (authData = {}, userAuthData, config) => {
+  if (!config.auth) return;
+
+  const missingRequiredProviders = Object.keys(config.auth).filter(
+    provider => config.auth[provider].required && !authData[provider]
+  );
+
+  if (missingRequiredProviders.length) {
+    throw new Parse.Error(
+      Parse.Error.OTHER_CAUSE,
+      `Missing required authData ${missingRequiredProviders.join(',')}`
+    );
+  }
+
+  // In case of signup we need to only check
+  // required providers
+  if (!userAuthData) return;
+
+  const savedUserProviders = Object.keys(userAuthData);
+
+  const hasProvidedASoloProvider = savedUserProviders.some(
+    provider => config.auth[provider] && config.auth[provider].policy === 'solo'
+  );
+
+  // Solo providers can be considered as safe
+  // so we do not have to check if the user need
+  // to provide an additional provider to login
+  if (hasProvidedASoloProvider) return;
+
+  const additionProvidersNotFound = [];
+  const hasProvidedAtLeastOneAdditionalProvider = savedUserProviders.some(provider => {
+    if (config.auth[provider] && config.auth[provider].policy === 'additional') {
+      if (authData[provider]) {
+        return true;
+      } else {
+        // Push missing provider for plausible error return
+        additionProvidersNotFound.push(provider);
+      }
+    }
+  });
+  if (hasProvidedAtLeastOneAdditionalProvider || !additionProvidersNotFound.length) return;
+
+  throw new Parse.Error(
+    Parse.Error.OTHER_CAUSE,
+    `Missing additional authData ${additionProvidersNotFound.join(',')}`
+  );
```

Comment:
Additional adpater handling (like mfa), here we are searching if an additional adpater was configured on the user